### PR TITLE
common/admin_socket: add new api unregister_commands(AdminSocketHook …

### PR DESCRIPTION
**Pull request by majianpeng**
…*hook)

We can unregister all commands which belong to this hook. Make code
looks clean and avoid forgeting unregister command.

Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

### CODE REVIEW

**Comment by majianpeng** [comment]
@tchaikov  . please review.

**Comment by majianpeng** [comment]
@tchaikov . ping

File: src/common/admin_socket.cc [hunk 1]
```diff
@@ -484,6 +484,27 @@ int AdminSocket::unregister_command(std::string_view command)
   return ret;
 }
 
+void AdminSocket::unregister_commands(const AdminSocketHook *hook)
+{
+  if (hook) {
```
Comment (tchaikov) [inline src/common/admin_socket.cc]: under what circumstances, we will have a `nullptr` hook?

Comment (majianpeng) [reply src/common/admin_socket.cc]: update.

File: src/common/admin_socket.cc [hunk 2]
```diff
@@ -484,6 +484,25 @@ int AdminSocket::unregister_command(std::string_view command)
   return ret;
 }
 
+void AdminSocket::unregister_commands(const AdminSocketHook *hook)
+{
+  std::unique_lock l(lock);
```
Comment (tchaikov) [inline src/common/admin_socket.cc]: nit, `std::scoped_lock` would suffice here.

File: src/common/admin_socket.cc [hunk 3]
```diff
@@ -484,6 +484,25 @@ int AdminSocket::unregister_command(std::string_view command)
...
+  auto i = hooks.begin();
+  while (i != hooks.end()) {
+    if (i->second.hook == hook) {
+      ldout(m_cct, 5) << "unregister_commands " << i->first << dendl;
```
Comment (tchaikov) [inline src/common/admin_socket.cc]: please use `__func__` instead if possible.

File: src/common/admin_socket.cc (hunk 2 above)
Comment (majianpeng) [reply src/common/admin_socket.cc]: std::scoped_lock is the same std::lock_guard.  There is a wait to need this lock.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-06-02_12:29:35-rados-wip-kefu-testing-2018-06-02-1754-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-06-02_16:38:39-rados-wip-kefu-testing-2018-06-02-1754-distro-basic-smithi/#

**Pull request merged**